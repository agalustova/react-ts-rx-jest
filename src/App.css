.App {
  text-align: center;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App--loading {
  background-color: #a00;
}

.characteristics {
  width: 30%;
}
.characteristic {
  display: flex;
}
.characteristic-name {
  width: 80%;
  display: flex;
  justify-content: flex-start;
}
