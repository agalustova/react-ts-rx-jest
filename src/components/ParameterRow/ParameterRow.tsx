import React from 'react';

interface ParametersRowProps {
  title: string,
  value: string | number,
}

function ParametersRow({
  title,
  value,
}: ParametersRowProps): React.ReactElement {
  return (
    <div className="characteristic">
      <div className="characteristic-name">{title}</div>
      <div>
        {value}
      </div>
    </div>
  );
}

export default ParametersRow;
