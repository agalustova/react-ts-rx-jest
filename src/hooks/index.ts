export { default as usePeriodicalValue } from './usePeriodicalValue';
export { default as useSensorsData } from './useSensorsData';
