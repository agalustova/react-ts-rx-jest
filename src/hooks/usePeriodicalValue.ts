import { useEffect, useState, useRef } from 'react';
import { range, of } from 'rxjs';
import { concatMap, delay } from 'rxjs/operators';

export enum SensorLabel {
  NOT_PREPARED = 'NOT_PREPARED',
  NOT_AVAILABLE_LABEL = 'N/A',
}

const {
  floor,
  random,
} = Math;

const RESET_TO_NOT_AVAILABLE_DELAY = 1000; // milliseconds
const MIN_INTERVAL_TIME = 100; // milliseconds
const MAX_INTERVAL_TIME = 2000; // milliseconds

function getRandomInteger(min = 0, max = 0) {
  return floor(min + random() * (max + 1 - min));
}

const getDelayTime = () => getRandomInteger(MIN_INTERVAL_TIME, MAX_INTERVAL_TIME);

function usePeriodicalValue() {
  const [value, setValue] = useState<number | string>(SensorLabel.NOT_PREPARED);
  const timer = useRef<any>(null);

  useEffect(() => {
    if (timer.current) {
      clearTimeout(timer.current);
    }
    if (value !== SensorLabel.NOT_AVAILABLE_LABEL && value !== SensorLabel.NOT_PREPARED) {
      timer.current = setTimeout(() => {
        timer.current = null;
        setValue(SensorLabel.NOT_AVAILABLE_LABEL);
      }, RESET_TO_NOT_AVAILABLE_DELAY);
    }

    return () => clearTimeout(timer.current);
  }, [value]);

  useEffect(() => {
    const obs = range(1, 400).pipe(
      concatMap((i) => of(i).pipe(
        delay(getDelayTime()),
      )),
    ).subscribe(() => {
      setValue(getRandomInteger(60, 70));
    });
    return () => obs.unsubscribe();
  }, []);

  return value;
}

export default usePeriodicalValue;
