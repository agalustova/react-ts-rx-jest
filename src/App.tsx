import React from 'react';
import clsx from 'clsx';
import { useSensorsData } from 'hooks';
import ParameterRow from 'components/ParameterRow';
import './App.css';

function App(): React.ReactElement {
  const {
    loading,
    temperature,
    pressure,
    humidity,
  } = useSensorsData();

  return (
    <div className="App">
      <header className={clsx('App-header', { 'App--loading': loading })}>
        {loading && (<div>Data will be available when sensors are prepared.</div>)}
        {!loading && (
          <div className="characteristics">
            <ParameterRow title="Temperature :" value={temperature} />
            <ParameterRow title="Air pressure :" value={pressure} />
            <ParameterRow title="Humidity :" value={humidity} />
          </div>
        )}
      </header>
    </div>
  );
}

export default App;
