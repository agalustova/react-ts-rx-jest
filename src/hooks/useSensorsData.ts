import { useEffect, useState } from 'react';
import usePeriodicalValue, { SensorLabel } from './usePeriodicalValue';

interface SensorsDataProps {
  loading: boolean;
  temperature: string | number;
  pressure: string | number;
  humidity: string | number;
}

// from REQUIREMENTS:
// All 3 systems must emit at least one value before 1 display object is ever sent to the dashboard.
function useSensorsData(): SensorsDataProps {
  const [allReady, setAllReady] = useState(false);
  const temperature = usePeriodicalValue();
  const pressure = usePeriodicalValue();
  const humidity = usePeriodicalValue();

  useEffect(() => {
    if (
      !allReady
      && temperature !== SensorLabel.NOT_PREPARED
      && pressure !== SensorLabel.NOT_PREPARED
      && humidity !== SensorLabel.NOT_PREPARED
    ) {
      setAllReady(true);
    }
  }, [temperature, pressure, humidity, allReady]);

  return {
    loading: !allReady,
    temperature,
    pressure,
    humidity,
  };
}

export default useSensorsData;
