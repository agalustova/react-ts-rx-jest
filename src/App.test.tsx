import React from 'react';
// import { render } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';
import ParameterRow from 'components/ParameterRow';

import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { useSensorsData } from 'hooks';
import App from './App';

import usePeriodicalValue, { SensorLabel } from './hooks/usePeriodicalValue';

Enzyme.configure({ adapter: new Adapter() });

// test('renders learn react link', () => {
//   const { getByText } = render(<App />);
//   const linkElement = getByText(/Data will be available when sensors are prepared./i);
//   expect(linkElement).toBeInTheDocument();
// });
describe('ParameterRow', () => {
  it('shallow ParameterRow Temperature component', () => {
    const myComponent = shallow(<ParameterRow title="Temperature :" value={68} />);
    expect(myComponent.find('div')).toHaveLength(3);
  });
  it('shallow ParameterRow Air component', () => {
    const myComponent = shallow(<ParameterRow title="Air pressure :" value={68} />);
    expect(myComponent.find('div')).toHaveLength(3);
  });
  it('shallow ParameterRow Humidity component', () => {
    const myComponent = shallow(<ParameterRow title="Humidity :" value={68} />);
    expect(myComponent.find('div')).toHaveLength(3);
  });
  it('renders properly', () => {
    const myComponent = shallow(<ParameterRow title="Humidity :" value={68} />);
    expect(myComponent).toMatchSnapshot();
  });
});
describe('App', () => {
  it('shallow App component', () => {
    const myComponent = shallow(<App />);
    expect(myComponent.find('div')).toHaveLength(2);
  });
  it('renders properly', () => {
    const myComponent = shallow(<App />);
    expect(myComponent).toMatchSnapshot();
  });
});

describe('get param initial value', () => {
  it('initial value', () => {
    const { result } = renderHook(() => usePeriodicalValue());

    expect(result.current).toBe('NOT_PREPARED');
  });
});
describe('get param value after 3 sec', () => {
  it('initial value', async () => {
    const { result, waitForNextUpdate } = renderHook(() => usePeriodicalValue());
    expect(result.current).toBe('NOT_PREPARED');
    await waitForNextUpdate();
    expect(result.current).toBeTruthy();
  });
});

describe('get all param initial value', () => {
  it('initial object value', () => {
    const { result } = renderHook(() => useSensorsData());
    const expectedResult = {
      humidity: 'NOT_PREPARED',
      loading: true,
      pressure: 'NOT_PREPARED',
      temperature: 'NOT_PREPARED',
    };
    expect(result.current).toStrictEqual(expectedResult);
  });
});
