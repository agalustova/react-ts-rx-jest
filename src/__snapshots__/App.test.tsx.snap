// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders properly 1`] = `
<div
  className="App"
>
  <header
    className="App-header App--loading"
  >
    <div>
      Data will be available when sensors are prepared.
    </div>
  </header>
</div>
`;

exports[`ParameterRow renders properly 1`] = `
<div
  className="characteristic"
>
  <div
    className="characteristic-name"
  >
    Humidity :
  </div>
  <div>
    68
  </div>
</div>
`;
